src/nupic/algorithms/Cells4.cpp

```diff
@@ -1574,6 +1575,44 @@ void Cells4::applyGlobalDecay()
 
 //--------------------------------------------------------------------------------
 /**
+ * Helper function for Cells4::adaptSegment. Generates lists of synapses to
+ * decrement, increment, add, and remove.
+ *
+ * IMPLEMENTATION NOTE: The implementation makes a single pass through the
+ * containers for O(n) complexity.
+ *
+ */
+void Cells4::_generateListsOfSynapsesToAdjustForAdaptSegment(
+  const Segment& segment,
+  std::set<UInt>& synapsesSet,
+  std::vector<UInt>& inactiveSrcCellIdxs,
+  std::vector<UInt>& inactiveSynapseIdxs,
+  std::vector<UInt>& activeSrcCellIdxs,
+  std::vector<UInt>& activeSynapseIdxs)
+{
+  // Purge residual data
+  inactiveSrcCellIdxs.clear();
+  inactiveSynapseIdxs.clear();
+  activeSrcCellIdxs.clear();
+  activeSynapseIdxs.clear();
+
+  for (UInt i = 0; i != segment.size(); ++i) {
+    UInt srcCellIdx = segment[i].srcCellIdx();
+    if (not_in(srcCellIdx, synapsesSet)) {
+      inactiveSrcCellIdxs.push_back(srcCellIdx);  // TODO: Check synapse still exists!
```

Comment:
@scottpurdy, I spent a bit more time looking into this. There are two parallel sources of information that represent synapses:
1. `InSynapse` objects inside Segments. Each `InSynapse` object contains a cell index value of the source cell that connects to this synapse.
2. `OutSynapse` objects in `cells4::_outSynapses`. Each `OutSynapse` contains the cell and segment indexes that the referring cell connects to.

The parallel info universes exist for optimization, apparently. There is also a method that's used for consistency checks between the two info universes: `Cells4::invariants`.

I examined the code where `_outSynapses` is modified, including the callers of such code, and concluded that `_outSynapses` is a "shadow" data structure, while the Synapses contained in Segment objects are the primary ground truth. At least, that's how it appears. The pattern that I've observe has been to add/remove synapses inside Segment(s) and then follow up with `addOutSynapses`, `eraseOutSynapses`, or `rebuildOutSynapses` to update the "shadow" `_outSynapses`.

If this assessment is correct, then it should be safe to assume that this TODO is misguided given the current state of the code. The synapses inside segments should be valid on entry to this code path and they are not modified by this block of code within this loop.

I am going to remove this TODO.